### PR TITLE
#41 Api Authentication

### PR DESCRIPTION
**Pull request by ovdmar**

### CODE REVIEW

File: testsuite/test_api.py [hunk 1]
```diff
@@ -19,7 +22,12 @@ def __init__(self, client):
         self.client = client
 
     def start(self):
-        resp = self.client.post('/v0/jobs')
+        credentials = base64.b64encode(b'test:test').decode('utf8')
+        resp = self.client.get('/v0/token', **{'HTTP_AUTHORIZATION': 'Basic ' + credentials}).json()
```
Comment (mgax) [inline testsuite/test_api.py]: You can save a few chars by just writing `resp = self.client.get('/v0/token', HTTP_AUTHORIZATION=f'Basic {credentials}').json()`

Comment (mgax) [reply testsuite/test_api.py]: Also please use a different variable name, e.g. `auth_resp`, makes reading the code easier.

File: vmck/api.py [hunk 2]
```diff
@@ -1,13 +1,19 @@
+import base64
 import json
+
```
Comment (mgax) [inline vmck/api.py]: No need for newline.

File: vmck/api.py [hunk 3]
```diff
@@ -1,13 +1,19 @@
...
-from django.views.decorators.http import require_http_methods
 from django.views.decorators.csrf import csrf_exempt
-from django.shortcuts import get_object_or_404
-from .backends import get_backend
+from django.views.decorators.http import require_http_methods
+
 from . import jobs
 from . import models
+from . import settings
+from .backends import get_backend
```
Comment (mgax) [inline vmck/api.py]: Ah, i like this import order better. Does your IDE do this?

File: vmck/api.py [hunk 4]
```diff
@@ -23,6 +29,28 @@ def home(request):
     })
 
 
+def create_token(request):
+    auth_header = request.headers['Authorization']
+    token_type, _, credentials = auth_header.partition(' ')
```
Comment (mgax) [inline vmck/api.py]: Please add parantheses around the variables, it makes it more clear that you're unpacking a list or tuple.

File: vmck/api.py [hunk 5]
```diff
@@ -23,6 +29,28 @@ def home(request):
...
+
+    username, password = base64.b64decode(credentials).decode('utf-8').split(':')
+    # TODO if password is not correct return failure
```
Comment (mgax) [inline vmck/api.py]: Please implement this too, with a test. We can simply return a `JsonResponse` with 401 status code and `{'error': "Invalid credentials"}` as body.

File: vmck/api.py [hunk 6]
```diff
@@ -23,6 +29,28 @@ def home(request):
...
+    private_key = settings.SECRET_KEY
+
+    jwt_token = jwt.encode(
+        {
+            'sub': username,
+            'iss': 'vmck',
+            'iat': int(time.time())
+        },
+        key=private_key,
+        algorithm='HS256'
```
Comment (mgax) [inline vmck/api.py]: Please add a final `,` as PEP8 recommends. Same for the dict above.

File: vmck/api.py [hunk 7]
```diff
@@ -57,6 +85,7 @@ def view(request, **kwargs):
 
 urls = [
     path('', route(GET=home)),
+    path('token', route(GET=create_token)),
     path('jobs', route(POST=create_job)),
     path('jobs/<int:pk>', route(GET=get_job, DELETE=kill_job)),
-]
+]
```
Comment (mgax) [inline vmck/api.py]: Please add a final newline.

File: vmck/middleware.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+                return HttpResponse('Unauthorized', status=401)
+
+            jwt_token = auth_header[7:]
+
+            private_key = settings.SECRET_KEY
+            decoded_jwt = jwt.decode(jwt_token, private_key, algorithms=['HS256'])
+            user = decoded_jwt['sub']
+
+            if user:
+                # TODO get user object and set it on request
```
Comment (mgax) [inline vmck/middleware.py]: What does this TODO mean? It looks like the next line does exactly it.

File: vmck/settings.py [hunk 9]
```diff
@@ -16,7 +16,7 @@
 DATABASES = {
     'default': {
         'ENGINE': 'django.db.backends.sqlite3',
-        'NAME': str(base_dir / 'data' / 'db.sqlite3'),
+        'NAME': os.path.join(base_dir, 'data', 'db.sqlite3'),
```
Comment (mgax) [inline vmck/settings.py]: No, please keep using `pathlib`, it's the new way of handling paths in Python.

File: vmck/api.py (hunk 4 above)
Comment (mgax) [reply vmck/api.py]: See django's api on how to verify user credentials: https://docs.djangoproject.com/en/2.2/topics/auth/default/#authenticating-users

**Comment by Jokeswar** [comment]
Not up to date.

**Pull request closed without merging**